always-hydrate.md

```diff
@@ -0,0 +1,133 @@
+following commands:
+
+```
+    gvfs prefetch
+      [<Repo Root Path>]
+      --list-always-hydrate
+```
+
+The `--list-always-hydrate` flag is incompatible with any of the other flags. It will read out the
+`always-hydrate` file and display the entries to the user.
+
+```
+    gvfs prefetch
+      [<Repo Root Path>]
+      [--files <semicolon-delimited list of file paths, prefix * wildcards allowed>] 
+      [--folders <semicolon-delimited list of folder paths, no wildcards>]
+      --remove-always-hydrate
+```
+
+The `--remove-always-hydrate` flag is incompatible with any combination of `--[always-]hydrate` or
+`--list-always-hydrate`.
+
+Any entries that match the specified `--files` and `--folders` arguments will be removed from the
+always-hydrate file. To remove all entries, the user can specify `gvfs prefetch 
+--remove-always-hydrate --files *`.
+
+## Implementation
+### Storing always-hydrate entries
+We will create a new file in the `.gvfs` folder called `always-hydrate`. This file is technically
+human readable and editable, but like all files in the `.gvfs` folder, carries no contract around
+maintaining the same format moving forward. All user interactions with this file should go through
+the `gvfs prefetch` command.
+
+Therefore the specific format of the file is not that interesting. For the sake of easy
+implementation and maintenance, we will reuse our existing file formats from our family of
+`FileBasedCollection` classes, either `FileBasedDictionary` or a new type that derives from
+`FileBasedCollection` in the pattern of the `PlaceholderListDatabase`.
+
+In this case, we need to store an entry per file or folder that was specified to the prefetch
+command, and we will need to remember if it was a file or a folder. Current plan is to go with a
+`FileBasedDictionary` unless we find the need to store more fields.
+
+### Hydrating files after a projection change
+Whenever the projection changes, the `GitIndexProjection` class currently kicks in and updates all
+placeholder files on disk with their new blob ids (for any files that changed). This is all done
+synchronously and blocks the git command that caused the projection change.
+
+In addition, this code will now also schedule background downloads of all paths that match the
+`always-hydrate` collection. We will just pass the list of file and folder parameters to the existing
+`BlobPrefetcher` that the `PrefetchVerb` uses to download and hydrate files today. 
+
+That class will determine which blobs need to be downloaded and which files need to be hydrated.
+However, the existing functionality of that class tries to spin up several threads and do bulk
+discovery, download, and hydration of files. Since we want these to be low-priority downloads (see 
+next section), we will instead just take the resulting list of blobs and paths, and schedule 
+individual requests for each one. (Note: some experimentation will be needed here. It may be more
```

Comment:
Let's think about this differently -- it will definitely be batched, but we can start with a batch size of 1. This makes it easier to experiment with later.

Reply:
Agreed